src/test/ui/consts/const-num-variants.rs

```diff
@@ -0,0 +1,36 @@
+// run-pass
+#![allow(dead_code)]
+#![feature(const_num_variants)]
+#![feature(num_variants)]
+
+use std::mem::num_variants;
+
+enum Void {}
+
+enum Foo {
+    A,
+    B,
+    C,
+}
+
+enum Bar {
+    A,
+    B,
+    C,
+    D(usize),
+    E { field_1: usize, field_2: Foo },
+}
+
+const TEST_VOID: usize = num_variants::<Void>();
+const TEST_FOO: usize = num_variants::<Foo>();
+const TEST_BAR: usize = num_variants::<Bar>();
+
+fn main() {
+    assert_eq!(TEST_VOID, 0);
```

Comment:
Could you add a test for a struct and some primitives (e.g `bool`, `*const ()`, `u8`)?

Reply:
As in to test there is no ICE, or to test for a specific return value?

Reply:
I'm not sure what the proper behavior should be, but we should have a test that verifies whatever we decide the correct behavior is.

Reply:
It's left as unspecified as with `core::mem::discriminant`

Reply:
The underlying intrinsic always returns zero at the moment